tests/torch/sparsity/movement/helpers.py

```diff
@@ -0,0 +1,189 @@
+
+    def compute_loss(self, model, inputs, return_outputs=False):
+        loss_main, outputs = super().compute_loss(model, inputs, return_outputs=True)
+        loss_compress = self.compression_ctrl.loss()
+        loss = loss_main + loss_compress
+        return (loss, outputs) if return_outputs else loss
+
+
+class BaseCallback(TrainerCallback):
+    def __init__(self, compression_ctrl: CompressionAlgorithmController) -> None:
+        self.compression_ctrl = compression_ctrl
+        self._compress_log = dict()
+
+    def get_compress_log(self):
+        return self._compress_log.copy()
+
+    def get_train_log(self):
+        return self._train_log.copy()
+
+    def on_train_end(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
+        self._train_log = state.log_history
+
+    def on_step_begin(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
+        self.compression_ctrl.scheduler.step()
+
+    def on_step_end(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
+        movement_ctrl_statistics = self.compression_ctrl.statistics().movement_sparsity
+        info = dict(
+            step=state.global_step,
+            epoch=state.epoch,
+            importance_regularization_factor=movement_ctrl_statistics.importance_regularization_factor,
+            importance_threshold=movement_ctrl_statistics.importance_threshold,
+            relative_sparsity=movement_ctrl_statistics.model_statistics.sparsity_level_for_layers,
+        )
+        self._compress_log[float(state.epoch)] = info
+
+    def on_epoch_begin(self, args: TrainingArguments, state: TrainerState, control: TrainerControl, **kwargs):
+        self.compression_ctrl.scheduler.epoch_step()
+        mvmt_ctrl = self.compression_ctrl  # TODO: mutliple child ctrls
+        # should move into the scheduler codes
+        if self.compression_ctrl.scheduler.get_state()["current_epoch"] == mvmt_ctrl.scheduler.warmup_end_epoch:
+            print('Fill stage')
+            mvmt_ctrl.reset_independent_structured_mask()
+            mvmt_ctrl.resolve_structured_mask()
+            mvmt_ctrl.populate_structured_mask()
+
+
+def run_movement_pipeline(tmp_path, compression_ctrl, compressed_model,
+                          callbacks: Optional[List[BaseCallback]] = None, **kwargs):  # temporarily remove fixture for model & dataset
+    train_dataset, eval_dataset = yelp_dataset()
+    args = dict(
+        output_dir=tmp_path / "test_trainer",
+        label_names=["labels"],
+        evaluation_strategy="epoch",
+        per_device_train_batch_size=32,
+        per_device_eval_batch_size=32,
+        num_train_epochs=6,
+        learning_rate=1e-3,
+        optim="adamw_torch",
+        remove_unused_columns=False,
+        report_to=None,
+        disable_tqdm=True,
+        no_cuda=True,  # TODO: where to set cuda devices for cuda training?
```

Comment:
We use `move_model_to_cuda_if_available(model)` function for setting device. Please keep in mind that sometimes there's no cuda available, and the test should work.
Will it work if we set device before launching pipeline, won't it?

Reply:
> We use `move_model_to_cuda_if_available(model)` function for setting device. Please keep in mind that sometimes there's no cuda available, and the test should work. Will it work if we set device before launching pipeline, won't it?

I find that many other tests are conducted on cpu only. Currently i will also focus on cpu since cuda is not a must in jpqd tests,.